arangod/Aql/LateMaterializedOptimizerRulesCommon.cpp

```diff
@@ -126,11 +132,123 @@ bool latematerialized::getReferencedAttributes(AstNode* node,
   return state.optimize;
 }
 
-bool latematerialized::isPrefix(std::vector<arangodb::basics::AttributeName> const& prefix,
+template<bool indexDataOnly, typename Attrs>
+bool attributesMatch(iresearch::IResearchViewSort const& primarySort,
+                     iresearch::IResearchViewStoredValues const& storedValues,
+                     Attrs& attrs,
+                     std::vector<std::vector<ColumnVariant<indexDataOnly>>>& usedColumnsCounter,
+                     size_t columnsCount) {
+  TRI_ASSERT(columnsCount <= usedColumnsCounter.size());
+  // check all node attributes to be in sort
+  std::remove_reference_t<decltype(usedColumnsCounter)> tmpUsedColumnsCounter(columnsCount);
+  typename ColumnVariant<indexDataOnly>::PostfixType postfix{};
+  for (auto& nodeAttr : attrs) {
+    auto found = false;
+    nodeAttr.afData.field = nullptr;
+    // try to find in the sort column
+    size_t fieldNum = 0;
+    TRI_ASSERT(!tmpUsedColumnsCounter.empty());
+    auto& tmpSlot = tmpUsedColumnsCounter.front();
+    for (auto const& field : primarySort.fields()) {
+      if (latematerialized::isPrefix<indexDataOnly>(field, nodeAttr.attr, false, postfix)) {
+        tmpSlot.emplace_back(&nodeAttr.afData, fieldNum, &field, std::move(postfix));
+        if constexpr (std::is_arithmetic_v<decltype(postfix)>) {
+          postfix = 0;
+        }
+        found = true;
+        break;
+      }
+      ++fieldNum;
+    }
+    // try to find in other columns
+    ptrdiff_t columnNum = 1;
+    for (auto const& column : storedValues.columns()) {
+      fieldNum = 0;
+      TRI_ASSERT(static_cast<ptrdiff_t>(tmpUsedColumnsCounter.size()) >= columnNum);
+      auto& tmpSlot = tmpUsedColumnsCounter[columnNum];
+      for (auto const& field : column.fields) {
+        if (latematerialized::isPrefix<indexDataOnly>(field.second, nodeAttr.attr, false, postfix)) {
+          tmpSlot.emplace_back(&nodeAttr.afData, fieldNum, &field.second, std::move(postfix));
+          if constexpr (std::is_arithmetic_v<decltype(postfix)>) {
+            postfix = 0;
+          }
+          found = true;
+          break;
+        }
+        ++fieldNum;
+      }
```

Comment:
The loop body looks identical to the loop body above (lines 152 to 162), just the input fields and `tmpSlot` different.
Is it possible to implement the loop body just once and call it from both places in this function?

Reply:
Fixed. Good catch.